# Importing Libraries and datasets products.csv and orders_wrangled.csv

In [1]:
import pandas as pd

In [2]:
import numpy as nyp

In [3]:
import os

In [5]:
path=r'/Users/stefaniadaskalaki/Data Immersion/4 Achievement Python/CF4.2 Instacart Basket Analysis'

In [6]:
df_prods=pd.read_csv(os.path.join(path, '02. Data', '01. Original Data', 'products.csv'), index_col=False)

In [7]:
df_ords=pd.read_csv(os.path.join(path, '02. Data', '02. Prepared Data', 'orders_wrangled.csv'), index_col=False)

# Create a dataframe

In [8]:
df_test=pd.DataFrame()

## Create a mixed type column

In [9]:
df_test['mix']=['a','b',1,True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

## Check for mixed type columns in df_test

In [12]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


## Fix column type from mix to string

In [13]:
df_test['mix'] = df_test['mix'].astype('str')

# Finding missing values in df_prods

In [14]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

# Creating df_nan for missing values from product_name

In [15]:
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## Compare missing values df and df_prodcuts

In [17]:
df_prods.shape

(49693, 5)

In [18]:
df_nan.shape

(16, 5)

since the percentage of missing values is way lower than 5% i choose to delete the 16 entries with product name missing value

# Create a new df_prods_clean excluding missing values

In [19]:
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [22]:
df_prods_clean.shape

(49677, 5)

# Finding full row duplicates

In [23]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Addressing duplicates

In [25]:
df_prods_clean.shape

(49677, 5)

In [26]:
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [27]:
df_prods_clean_no_dups.shape

(49672, 5)

Here i checked how many rows our initial df_prods_clean had then deducted the duplicates and created a new df_prods_clean_no_dups without them. The result shows that correctly 5 rows have been deleted.

# Export df_prods_clean_no_dups

In [28]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02. Data','02. Prepared Data', 'products_checked.csv'))

# TASK 4.5 

## 2.Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [29]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The max value of product_id is 49688 which is 5 short from the total count of 49693. Additionally the max value for price is 99999 which is possibly a mistake. 

In [31]:
df_prods.tail(10)

product_id                                product_name  aisle_id  \
49683       49679                     Famous Chocolate Wafers        61   
49684       49680          All Natural Creamy Caesar Dressing        89   
49685       49681     Spaghetti with Meatballs and Sauce Meal        38   
49686       49682                          California Limeade        98   
49687       49683                              Cucumber Kirby        83   
49688       49684   Vodka, Triple Distilled, Twist of Vanilla       124   
49689       49685          En Croute Roast Hazelnut Cranberry        42   
49690       49686                            Artisan Baguette       112   
49691       49687  Smartblend Healthy Metabolism Dry Cat Food        41   
49692       49688                      Fresh Foaming Cleanser        73   

       department_id  prices  
49683             19     6.0  
49684             13     4.9  
49685              1     6.9  
49686              7     4.3  
49687              4    13.2  
49688              5     5.3  
49689              1     3.1  
49690              3     7.8  
49691              8     4.7  
49692             11    13.5

In [32]:
df_prods[df_prods.duplicated()]

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

By performing a .tail(10) i can see that there is a discrepancy between product_id and index which means there must be duplicates and this confirms our findings from earlier and again proving here by performing the duplicates function.

In [33]:
df_prods[df_prods.prices==99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

By checking which product item has the 99999 price tag proves that it is definitely an error as milk could not cost 99999.

# 3.Check for mixed-type data in your df_ords dataframe.

In [36]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

since we don't get any result it means that all columns are single data type.

# 5.Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [37]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [42]:
df_ords.head(30)

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0            0   2539329        1             1                  2   
1            1   2398795        1             2                  3   
2            2    473747        1             3                  3   
3            3   2254736        1             4                  4   
4            4    431534        1             5                  4   
5            5   3367565        1             6                  2   
6            6    550135        1             7                  1   
7            7   3108588        1             8                  1   
8            8   2295261        1             9                  1   
9            9   2550362        1            10                  4   
10          10   1187899        1            11                  4   
11          11   2168274        2             1                  2   
12          12   1501582        2             2                  5   
13          13   1901567        2             3                  1   
14          14    738281        2             4                  2   
15          15   1673511        2             5                  3   
16          16   1199898        2             6                  2   
17          17   3194192        2             7                  2   
18          18    788338        2             8                  1   
19          19   1718559        2             9                  2   
20          20   1447487        2            10                  1   
21          21   1402090        2            11                  1   
22          22   3186735        2            12                  1   
23          23   3268552        2            13                  4   
24          24    839880        2            14                  3   
25          25   1492625        2            15                  1   
26          26   1374495        3             1                  1   
27          27    444309        3             2                  3   
28          28   3002854        3             3                  3   
29          29   2037211        3             4                  2   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0  
11                 11                     NaN  
12                 10                    10.0  
13                 10                     3.0  
14                 10                     8.0  
15                 11                     8.0  
16                  9                    13.0  
17                 12                    14.0  
18                 15                    27.0  
19                  9                     8.0  
20                 11                     6.0  
21                 10                    30.0  
22                  9                    28.0  
23                 11                    30.0  
24                 10                    13.0  
25                 11                    30.0  
26                 14                     NaN  
27                 19                     9.0  
28                 16                    21.0  
29                 18                    20.0

The only column with missing values is the days_since_prior_order with 206209 missing ones. I would think that these entries are missing because there are customers who purchased only once through the website. Additionally after I ran a check for the frist 30 entries in the df_ords I notice that every first order of each customer is marked as NaN in the days_since_prior_order column.

# 6.Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

In [44]:
df_ords_clean=df_ords

In [45]:
df_ords_clean['new_customer']=df_ords['days_since_prior_order'].isnull()==True

In [46]:
df_ords_clean.head(30)

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0            0   2539329        1             1                  2   
1            1   2398795        1             2                  3   
2            2    473747        1             3                  3   
3            3   2254736        1             4                  4   
4            4    431534        1             5                  4   
5            5   3367565        1             6                  2   
6            6    550135        1             7                  1   
7            7   3108588        1             8                  1   
8            8   2295261        1             9                  1   
9            9   2550362        1            10                  4   
10          10   1187899        1            11                  4   
11          11   2168274        2             1                  2   
12          12   1501582        2             2                  5   
13          13   1901567        2             3                  1   
14          14    738281        2             4                  2   
15          15   1673511        2             5                  3   
16          16   1199898        2             6                  2   
17          17   3194192        2             7                  2   
18          18    788338        2             8                  1   
19          19   1718559        2             9                  2   
20          20   1447487        2            10                  1   
21          21   1402090        2            11                  1   
22          22   3186735        2            12                  1   
23          23   3268552        2            13                  4   
24          24    839880        2            14                  3   
25          25   1492625        2            15                  1   
26          26   1374495        3             1                  1   
27          27    444309        3             2                  3   
28          28   3002854        3             3                  3   
29          29   2037211        3             4                  2   

    order_hour_of_day  days_since_prior_order  new_customer  
0                   8                     NaN          True  
1                   7                    15.0         False  
2                  12                    21.0         False  
3                   7                    29.0         False  
4                  15                    28.0         False  
5                   7                    19.0         False  
6                   9                    20.0         False  
7                  14                    14.0         False  
8                  16                     0.0         False  
9                   8                    30.0         False  
10                  8                    14.0         False  
11                 11                     NaN          True  
12                 10                    10.0         False  
13                 10                     3.0         False  
14                 10                     8.0         False  
15                 11                     8.0         False  
16                  9                    13.0         False  
17                 12                    14.0         False  
18                 15                    27.0         False  
19                  9                     8.0         False  
20                 11                     6.0         False  
21                 10                    30.0         False  
22                  9                    28.0         False  
23                 11                    30.0         False  
24                 10                    13.0         False  
25                 11                    30.0         False  
26                 14                     NaN          True  
27                 19                     9.0         False  
28                 16              

I created a new column in a new df_ords_cleam that will flag all the NaN values as 'New Customer' df_ords_clean. In this way we will be able to count/track all new customers without loosing any entries.

# 7.Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [47]:
df_ords_clean[df_ords_clean.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

There are no duplicates in this df_ords

# 9.Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [48]:
df_ords_clean.to_csv(os.path.join(path, '02. Data','02. Prepared Data', 'orders_checked.csv'))